Refactor `_resampled_scene()` and `_reduce_data()` methods of the `Scene` class

pnuu opened the pull request:
This PR refactors `Scene._resampled_scene()` that @gerritholl reported to be complicated in https://github.com/pytroll/satpy/pull/3168#issuecomment-3109062005 . As there was a change to `_reduce_data()`, I did some additional refactoring to it too.

 - [ ] Closes #xxxx <!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - [ ] Tests added <!-- for all bug fixes or enhancements -->
 - [ ] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->
 - [ ] Add your name to `AUTHORS.md` if not there already

pnuu committed:
Refactor Scene._resampled_scene()

pnuu committed:
Move source area access to _reduce_data()

pnuu committed:
Refactor data reduction

djhoese commented on satpy/scene.py:
```diff
@@ -878,34 +876,59 @@ def _resampled_scene(self, new_scn, destination_area, reduce_data=True,
+    @classmethod
+    def _get_new_datasets_from_parent(self, new_datasets, parent_dataset):
+        if parent_dataset is not None:
+            return new_datasets[DataID.from_dataarray(parent_dataset)]
```
`DataID.from_dataarray` returns a single DataID, right? I think I'd prefer a different name for this method. I think the purpose of this chunk of code is to say "if we've resampled the parent already, use the resampled parent", right? Or rather, if the current dataset has a parent, it should have been resampled already, so we should use the resampled version of the parent. I think?

In addition to renaming, it seems that `parent_dataset` is only used in the later steps to check for `is None`. I'm wondering if we can remove the use of `parent_dataset` in favor of `pres` and "bundle" this methods operation with the `dataset_walker` to be something like:

```python
for ds_id, dataset, resampled_parent in resampled_dataset_walker(datasets, new_datasets):
```

Or something like that.

...and if that is done, then there might be an argument for putting `_replace_anc_for_new_datasets` and `_update_area` into the for loop generator too. This changes the purpose of the for loop to be "what datasets do we need to resample" and then the inside of the for loop logic is just "reduce data", "resample data", "store result".

I'll admit the code was ugly and the logic of `new_scn._datasets` and `new_datasets` really isn't helping that.

djhoese replied:
Sorry for all the comments and no regular review. I just keep brainstorming.

One other idea, what if only `new_datasets` gets modified in the for loop and then assigning to `new_scn._datasets` is left for a second for loop (ex. `for ds_id, new_data_arr in new_datasets.items():`)? Maybe that would clean up the code inside the loop. I feel like a lot of this ugliness is caused by new_scn (or rather the `DatasetDict` inside) copying the DataArray and/or making modifications to it.

pnuu commented on satpy/scene.py:
```diff
@@ -878,34 +876,59 @@ def _resampled_scene(self, new_scn, destination_area, reduce_data=True,
                 replace_anc(res, pres)
 
+    @classmethod
+    def _get_new_datasets_from_parent(self, new_datasets, parent_dataset):
```
I'll have a look, thanks.

pnuu commented on satpy/scene.py:
```diff
@@ -927,32 +950,39 @@ def _prepare_resampler(self, source_area, destination_area, resamplers, resample
             resamplers[source_area] = resampler
             self._resamplers[key] = resampler
 
-    def _reduce_data(self, dataset, source_area, destination_area, reduce_data, reductions, resample_kwargs):
+    def _reduce_data(self, dataset, destination_area, reduce_data, reductions, resample_kwargs):
+        source_area = dataset.attrs["area"]
+        if not reduce_data:
+            LOG.debug("Data reduction disabled by the user")
+            return dataset, source_area
+
         try:
-            if reduce_data:
-                key = source_area
-                try:
-                    (slice_x, slice_y), source_area = reductions[key]
-                except KeyError:
-                    if resample_kwargs.get("resampler") == "gradient_search":
-                        factor = resample_kwargs.get("shape_divisible_by", 2)
-                    else:
-                        factor = None
-                    try:
-                        slice_x, slice_y = source_area.get_area_slices(
-                            destination_area, shape_divisible_by=factor)
-                    except TypeError:
-                        slice_x, slice_y = source_area.get_area_slices(
-                            destination_area)
-                    source_area = source_area[slice_y, slice_x]
-                    reductions[key] = (slice_x, slice_y), source_area
-                dataset = self._slice_data(source_area, (slice_x, slice_y), dataset)
-            else:
-                LOG.debug("Data reduction disabled by the user")
+            slice_x, slice_y = self._get_source_dest_slices(source_area, destination_area, reductions, resample_kwargs)
+            source_area = source_area[slice_y, slice_x]
+            reductions[source_area] = (slice_x, slice_y), source_area
+            dataset = self._slice_data(source_area, (slice_x, slice_y), dataset)
```
`NotImplementedError` or `KeyError`?

pnuu replied:
Nevermind, the try/except was just outside of this snippet. I'll see what can trigger the error.

lahtinep committed:
Move class methods to functions

lahtinep committed:
Move unnecessary code out of try block

lahtinep committed:
Use dataclass to hold resampling context

pnuu commented:
I made the Fine Choice(tm) of asking Copilot (Claude Sonnet 4.6) to refactor the method. After couple of iterations we ended up with this version that uses a data class to hold resampling context and a method that is called for each resampled dataset.

djhoese commented:
Interesting...I've never personally used data classes beyond toy programs and watching programming videos. What it is doing here makes me wonder if this deserves a full blown helper *class* where instead of `context` in these sub-methods it is just `self.`.

djhoese commented:
And sorry, I don't fully remember all the previous context for this PR and past thoughts I had about it.

lahtinep committed:
Merge branch 'main' into refactor_resampled_scene

lahtinep committed:
Merge branch 'main' into refactor_resampled_scene

pnuu commented:
Unrelated test failure:

```
ERROR satpy/tests/reader_tests/test_mviri_l1b_fiduceo_nc.py::TestFiduceoMviriFileHandlers - satpy/tests/reader_tests/test_mviri_l1b_fiduceo_nc.py::TestFiduceoMviriFileHandlers::test_init: duplicate parametrization of 'projection_longitude'
```

I can't replicate this locally with package versions I have.

pnuu commented:
Seems to be somehow related to https://github.com/pytest-dev/pytest/issues/13974 that have several links from the few last days.

pnuu commented:
Possible fix in https://github.com/pytroll/satpy/pull/3415

gerritholl commented:
Needs more test coverage?

gerritholl commented:
This PR increases the overhead/runtime of `_reduce_data` by a factor 5.  Test script:

```
import os
from satpy import Scene
from glob import glob
from satpy.tests.utils import CustomScheduler
from satpy.writers.core.compute import compute_writer_results
import dask.config
import time
reduce = True
with dask.config.set(scheduler=CustomScheduler(max_computes=0)):
    seviri_files = glob("/media/nas/x21308/scratch/SEVIRI/202103300900/H-000*")
    sc = Scene(filenames={"seviri_l1b_hrit": seviri_files})
    nms = {"ash", "natural_color", "dwd_RGB_12_12_9i_N", "convection",
           "dwd_IR_108", "dwd_HRV", "airmass", "dust"}
    sc.load(nms, generate=False)
    tic = time.time()
    ls = sc.resample("nqeuro3km",
                     resampler="gradient_search", reduce_data=reduce)
    print("resample", time.time() - tic, "seconds")
    res_all = []
    for nm in nms:
        res_all.append(ls.save_dataset(
                nm,
                "{platform_name}-{sensor}-{name}-{area.area_id}-{start_time:%Y%m%d%H%M}-geotiff.tif",
                writer="ninjogeotiff",
                ChannelID="IR 10.8",
                DataType="GORN",
                PhysicValue="Temperature",
                PhysicUnit="K",
                SatelliteNameID="METEOSAT Europe-Atlantic",
                fill_value=0,
                #dynamic_fields={"mean_time"},
                compute=False))
```

Satpy main: 8.2 seconds

This PR: 44.0 seconds

gerritholl commented:
Profiling reveals:

* the time per call for `_reduce_data` increases from 1.02 to 5.97 seconds.
* The number of calls to `spherical.py:492(get_next_intersection)` increases from 302 to 1812.
* The number of calls to `geometry.py:2654(get_area_slices)` (first entry point into pyresample) increases from 2 to 12.

Here both profiles:

[profile-results.zip](https://github.com/user-attachments/files/29285455/profile-results.zip)

gerritholl commented on satpy/scene.py:
```diff
@@ -927,32 +963,40 @@ def _prepare_resampler(self, source_area, destination_area, resamplers, resample
+    @staticmethod
+    def _get_source_dest_slices(source_area, destination_area, reductions, resample_kwargs):
+        try:
+            (slice_x, slice_y), source_area = reductions[source_area]
```
Something is going wrong with the reductions dictionary.  With satpy main, my test scripts goes twice into the `except KeyError` block, but the remaining 10 times it finds the reductions in the dictionary.  With this PR, it never finds an entry in the reductions dictionary, going 12 times into the except block.

gerritholl replied:
Fixed in f70d80f2c

gerritholl commented on satpy/scene.py:
```diff
@@ -66,6 +67,19 @@ class DelayedGeneration(KeyError):
+    resample_kwargs: dict
+    resampled_datasets: dict = dataclasses.field(default_factory=dict)
+    resamplers: dict = dataclasses.field(default_factory=dict)
+    reductions: dict = dataclasses.field(default_factory=dict)
```
I wonder if moving this into a dataclass breaks something?  In any case, the reductions dictionary has stopped working.  I haven't looked deeper yet.

gerritholl replied:
No, that was not the problem.

gerritholl reviewed (changes requested):
Serious performance regression in `_reduce_data`.  The `reductions` dictionary appears no longer functional, leading to `_reduce_data` being called for every dataset.  For my test script, that increases the time for calling `resample` from ca. 8 seconds to ca. 40 seconds.  Without computing anything.

gerritholl commented on satpy/scene.py:
```diff
@@ -927,32 +963,40 @@ def _prepare_resampler(self, source_area, destination_area, resamplers, resample
             LOG.info("Not reducing data before resampling.")
+        else:
+            source_area = source_area[slice_y, slice_x]
+            reductions[source_area] = (slice_x, slice_y), source_area
```
This is wrong.  The reductions dictionary should contain `source_area`, not `source_area[slice_y, slice_x]`.  Previously this was retained in `key = source_area` but this retaining got destroyed in the refactoring.

gerritholl commented on satpy/scene.py:
```diff
@@ -927,32 +963,40 @@ def _prepare_resampler(self, source_area, destination_area, resamplers, resample
-                LOG.debug("Data reduction disabled by the user")
+            slice_x, slice_y = self._get_source_dest_slices(source_area, destination_area, reductions, resample_kwargs)
         except NotImplementedError:
             LOG.info("Not reducing data before resampling.")
+        else:
+            source_area = source_area[slice_y, slice_x]
+            reductions[source_area] = (slice_x, slice_y), source_area
+            dataset = self._slice_data(source_area, (slice_x, slice_y), dataset)
```
This would seem to fix the performance regression:

```suggestion
        else:
            orig_source_area = source_area
            source_area = source_area[slice_y, slice_x]
            reductions[orig_source_area] = (slice_x, slice_y), source_area
            dataset = self._slice_data(source_area, (slice_x, slice_y), dataset)
```

gerritholl replied:
Fixed in 70d80f

gerritholl committed:
Fix performance regression

Fix a performance regression in _reduce_data, where the reductions
dictionary was no longer used correctly, leading to reduce_data being
called for every dataset instead of every source area.
